cpp/lib/ResultSetArrow.hpp

```diff
@@ -0,0 +1,255 @@
+     * @return 0 if successful, otherwise an error is returned.
+     */
+    SF_STATUS STDCALL getCurrCellAsUint64(uint64 * out_data);
+
+    /**
+     * Writes the value of the current cell as a float64 to the provided buffer.
+     *
+     * @param out_data             The buffer to write to.
+     *
+     * @return 0 if successful, otherwise an error is returned.
+     */
+    SF_STATUS STDCALL getCurrCellAsFloat64(float64 * out_data);
+
+    /**
+     * Writes the value of the current cell as a constant C-string to the provided buffer.
+     *
+     * @param out_data             The buffer to write to.
+     *
+     * @return 0 if successful, otherwise an error is returned.
+     */
+    SF_STATUS STDCALL getCurrCellAsConstString(const char ** out_data);
+
+    /**
+     * Writes the value of the current cell as a C-string to the provided buffer.
+     *
+     * In the event that the provided buffer is not large enough to contain the requested string,
+     * the buffer will be re-allocated and io_capacity will be updated accordingly.
+     *
+     * @param out_data             The buffer to write to.
+     * @param io_len               The length of the string.
+     * @param io_capacity          The capacity of the provided buffer.
+     *
+     * @return 0 if successful, otherwise an error is returned.
+     */
+    SF_STATUS STDCALL getCurrCellAsString(char * out_data, size_t * io_len, size_t * io_capacity);
+
+    /**
+     * Writes the value of the current cell as a timestamp to the provided buffer.
+     *
+     * @param out_data             The buffer to write to.
+     *
+     * @return 0 if successful, otherwise an error is returned.
+     */
+    SF_STATUS STDCALL getCurrCellAsTimestamp(SF_TIMESTAMP * out_data);
+
+    /**
+     * Gets the total number of rows in the current chunk being processed.
+     *
+     * @return the total number of rows in the current chunk.
+     */
+    size_t getRowCountInChunk();
+
+private:
+
+    /**
+     * Helper method to initialize the result set data.
+     *
+     * Populates m_records with the results in the first chunk, passed in as an
+     * argument to the constructor.
+     *
+     * @param initialChunk         The first chunk of the result set.
+     */
+    void init(cJSON * initialChunk);
+
+    /**
+     * BufferBuilder object used to consume chunks as they are retrieved from
+     * the server and continually build the final columnar result set.
+     */
+    std::shared_ptr<arrow::BufferBuilder> m_bufferBuilder;
+
+    /**
+     * BufferReader object used to read the chunk content in m_bufferBuilder.
+     * It is used in conjunction with m_recordBatchReader to iterate over the
+     * batches retrieved from the server to build the final columnar result set.
+     */
+    std::shared_ptr<arrow::io::BufferReader> m_bufferReader;
+
+    /**
+     * RecordBatchReader object used to operate on chunk content.
+     */
+    std::shared_ptr<arrow::ipc::RecordBatchReader> m_batchReader;
+
```

Comment:
These can be temporary variables and may not be required, If you are reusing the ResultSetArrow object then it may makes sense to have these variables and in that case these need to be reset after converting this buffer into vector of record batches and passed to arrowChunkIterator.

Reply:
Thanks, I agree these aren't necessarily required.

My refactor of `client.c` is such that the result set object gets destroyed when `_snowflake_stmt_reset()` is invoked. Since this happens for every call to `_snowflake_internal_query()`, which gets used in transaction calls, and `snowflake_prepare()` don't think that we will be reusing the same object to rebuild from streams. I've made these temporary variables instead. Let me know if this assessment is wrong.